reps/2023-02-03-node-affinity-feature-enhancements.md

```diff
@@ -0,0 +1,399 @@
+        scheduling_strategy=node_affinity([
+            label_in("region", ["china"], is_soft=false),
+            label_in("zone", ["china-east-1"], is_soft=true)
+        ])
+    ).remote()
+```
+**5. NodeAffinity can be used together with the CustomResource mechanism.**  
+These two mechanisms are completely non-conflicting.  
+```
+actor_1 = Actor.options(
+                num_gpus=1,
+                resources={"custom_resource": 1},
+                scheduling_strategy=node_affinity([
+                    label_exist("gpu_type", is_soft=false),
+                    label_in("gpu_type", ["A100", "P100"], is_soft=true),
+                ]).remote()
+```
+
+**FAQ:**  
+This node static labels seems to be very similar to custom_resources. Why not use custom_resources to instance of node_affinity?
+
+1. The semantics of static labels and custom_resources are completely different. custom_resources is a countable resource for nodes. And labels is an attribute or identifier of a node.
+2. Now custom_resources cannot implement the function of anti-affinity.
+3. Using labels to classify nodes and implement affinity/anti-affinity will be more natural and easier for users to accept.
+
+### Implementation plan
+
+![NodeAffinity Concept](https://user-images.githubusercontent.com/11072802/217133358-19ddc916-bf15-4f69-96e1-9cbd157b0e1d.png)
+
+1. Add the labels to node info
+```
+message GcsNodeInfo {
+  // The ID of node.
+  bytes node_id = 1;
+  ...
+  // The user-provided identifier or name for this node.
+  string node_name = 12;
+  // The static labels of nodes.
+  map<String, String> labels = 13;
+}
+```
+2. Add the labels data structure to the resource synchronization data structure(NodeResources). 
+```
+NodeResources {
+  ResourceRequest total;
+  ResourceRequest available;
+  /// Only used by light resource report.
+  ResourceRequest load;
+  /// Resources owned by normal tasks.
+  ResourceRequest normal_task_resources
+  /// Map<label_type, Map<namespace, Map<label_key, Map<label_value, ref_count>>>> Actors/Tasks/Nodes labels information
+  absl::flat_hash_map<string, absl::flat_hash_map<string, absl::flat_hash_map<string, absl::flat_hash_map<string, int>>>> labels;
+}
+```
+
+3. When a node registers with GCS and broadcasts to all nodes, add the static labels information to NodeResources.  
+
+Because NodeInfo originally had a mechanism to synchronize to the entire cluster. So there is no need to modify too much code.
+
+4. Scheduling optimization through Labels 
+
+Now any node raylet has node static labels information for all nodes.  
+when NodeAffinity schedules, if it traverses the Labels of each node, the algorithm complexity is very large, and the performance will be poor.   
+<b> Therefore, it is necessary to generate a full-cluster node labels index table to improve scheduling performance. </b>
+
+The label index table is only used to obtain which nodes have this key and value. The complexity of this algorithm is O(1).
```

Comment:
If there are multiple expressions to match, I think it will be O(kN) (k = num expressions, N = nodes), though, right?

Reply:
Yes, In a scene with many expressions, each time an expression is processed, fewer and fewer nodes meet the requirements. Performance is good.